stdlib/runtime/numbers.gr

```diff
@@ -1585,29 +1585,46 @@ let i64abs = x => {
   from WasmI64 use { (-), (>=) }
   if (x >= 0N) x else 0N - x
 }
-
 @unsafe
 let numberMod = (x, y) => {
   from WasmI64 use { (!=), (-), (*), (<), (>) }
   // incRef x and y to reuse them via WasmI32.toGrain
   Memory.incRef(x)
   Memory.incRef(y)
-  let xval = coerceNumberToWasmI64(WasmI32.toGrain(x): Number)
-  let yval = coerceNumberToWasmI64(WasmI32.toGrain(y): Number)
-  if (WasmI64.eqz(yval)) {
-    throw Exception.ModuloByZero
-  }
-  // We implement true modulo
-  if (xval < 0N && yval > 0N || xval > 0N && yval < 0N) {
-    let modval = WasmI64.remS(i64abs(xval), i64abs(yval))
-    let result = if (modval != 0N) {
-      i64abs(yval) - modval * (if (yval < 0N) -1N else 1N)
+  if (isFloat(x) || isFloat(y) || isRational(x) || isRational(y)) {
+    from WasmF64 use { (==), (/), (*), (-) }
+    let xval = coerceNumberToWasmF64(WasmI32.toGrain(x): Number)
+    let yval = coerceNumberToWasmF64(WasmI32.toGrain(y): Number)
+    let yInfinite = yval == InfinityW || yval == -InfinityW
+    let out = if (
+      yval == 0.0W || yInfinite && (xval == InfinityW || xval == -InfinityW)
+    ) {
+      newFloat64(NaNW)
+    } else if (yInfinite) {
+      newFloat64(xval)
     } else {
-      modval
+      newFloat64(xval - WasmF64.trunc(xval / yval) * yval)
     }
-    reducedInteger(result)
+    Memory.incRef(out)
```

Comment:
Do we need this `incRef`?

Reply:
When i didnt have it i was getting `decRef` errors. so i am pretty sure we need it.

Reply:
@ospencer Does that make sense to you? I'm a bit confused, since it looks like `reducedInteger` doesn't require an `incRef`. It's been a minute since I grokked this, though.

Reply:
No, this shouldn't be necessary. If there are decRef errors, the bug is somewhere else, possibly where this function is being called.

Reply:
Would someone who understands how the gc works a bit better be willing to take a look into why there is a `decRef` error when this doesnt exist?

Reply:
@ospencer can you please help Jake debug this so we can land it?

Reply:
Way ahead of you